common/src/main/java/io/netty/util/internal/PlatformDependent.java

```diff
@@ -1018,51 +1019,48 @@ private static boolean isRoot0() {
         return false;
     }
 
-    @SuppressWarnings("LoopStatementThatDoesntLoop")
     private static int javaVersion0() {
-        int javaVersion;
+        final int majorVersion;
 
-        // Not really a loop
-        for (;;) {
-            // Android
-            if (isAndroid()) {
-                javaVersion = 6;
-                break;
-            }
+        if (isAndroid()) {
+            majorVersion = 6;
+        } else {
+            majorVersion = majorVersionFromJavaSpecificationVersion();
+        }
 
-            try {
-                Method getVersion = java.lang.Runtime.class.getMethod("version");
-                Object version = getVersion.invoke(null);
-                javaVersion = (Integer) version.getClass().getMethod("major").invoke(version);
-                break;
-            } catch (Throwable ignored) {
-                // Ignore
-            }
+        logger.debug("Java version: {}", majorVersion);
 
-            try {
-                Class.forName("java.time.Clock", false, getClassLoader(Object.class));
-                javaVersion = 8;
-                break;
-            } catch (Throwable ignored) {
-                // Ignore
-            }
+        return majorVersion;
+    }
 
-            try {
-                Class.forName("java.util.concurrent.LinkedTransferQueue", false, getClassLoader(BlockingQueue.class));
-                javaVersion = 7;
-                break;
-            } catch (Throwable ignored) {
-                // Ignore
-            }
+    static int majorVersionFromJavaSpecificationVersion() {
+        try {
+            final String javaSpecVersion = AccessController.doPrivileged(new PrivilegedAction<String>() {
+                @Override
+                public String run() {
+                    return System.getProperty("java.specification.version");
+                }
+            });
+            return majorVersion(javaSpecVersion);
```

Comment:
@normanmaurer I do not think so, as it appears to me that [`java.specification.version` is required by the Java API spec](https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#getProperties--)?